pcicrawler/ocp_diag.py

```diff
@@ -0,0 +1,259 @@
+#!/usr/bin/env python3
+
+"""
+Copyright (c) Meta Platforms, Inc. and affiliates.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+"""
+
+import json
+import socket
+import pkg_resources
+import typing as ty
+import ocptv.output as tv
+from pci_lib import PCIDevice
+from abc import abstractmethod
+
+
+class OCPTestStep:
+    def __init__(self,
+                    device: PCIDevice,
+                    expected_result: str,
+                    ocp_run: tv.TestRun):
+        self._device = device
+        self._expected_result = expected_result
+        self._ocp_run = ocp_run
+        self._error_messages = []
+
+    @property
+    @abstractmethod
+    def _test_name(self):
+        pass
+
+    @abstractmethod
+    def run(self, current_value: str):
+        step = self._ocp_run.add_step(self._test_name)
+        if str(current_value) != str(self._expected_result):
+            step.add_diagnosis(tv.DiagnosisType.FAIL, 
+                            verdict="Test {} for device {}, expected {}, found {}.".format(
+                                self._test_name,
+                                self._device,
+                                self._expected_result,
+                                str(current_value)))
+            return False
+        else:
+            step.add_diagnosis(tv.DiagnosisType.PASS, 
+                            verdict="Test {} for device {} PASSED".format(
+                                self._test_name,
+                                self._device))
+            return True
+
+
+class CheckLocation(OCPTestStep):
+    _test_name = "pci_location_check"
+
+    def run(self):
+        return super().run(self._device.location)
+
+
+class CheckVendorID(OCPTestStep):
+    _test_name = "pci_vendor_id_check"
+
+    def run(self):
+        return super().run(self._device.vendor_id)
+
+
+class CheckDeviceID(OCPTestStep):
+    _test_name = "pci_device_id_check"
+
+    def run(self):
+        return super().run(self._device.device_id)
+
+
+class CheckClassID(OCPTestStep):
+    _test_name = "pci_class_id_check"
+
+    def run(self):
+        return super().run(self._device.class_id)
+
+
+class CheckPhysicalSlot(OCPTestStep):
+    _test_name = "pci_physical_slot_check"
+
+    def run(self):
+        return super().run(self._device.express_slot.slot)
```

Comment:
self._device.express_slot is not always existing.

Reply:
good catch; though i would rephrase this as `self._device.express_slot.slot` does not always exist, by virtue of `express_slot` being none

the same thing is a problem with `_device.express_link` in other classes in this file.

on a sidenote, this is actually a design problem in the `PCIDevice` class. The semantic is incorrect if it sometimes is a pci device and other times it's a pcie device (which is similar to a Liskov substitution problem). regardless, currently the code here may just throw, which needs to be fixed

Reply:
> self._device.express_slot is not always existing.

I'm not sure how we approach this then, should we error out or should we bypass? Both are reasonable depending on how we want to deal with this.